experimental/incremental/task.go

```diff
@@ -0,0 +1,234 @@
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package incremental
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Task represents a query that is currently being executed.
+//
+// Values of type Task are passed to [Query]. The main use of a Task is to
+// be passed to [Resolve] to resolve dependencies.
+type Task struct {
+	// We need all of the contexts for a call to [Run] to be the same, so to
+	// avoid user implementations of Query making this mistake (or inserting
+	// inappropriate timeouts), we pass the context as part of the task context.
+	ctx    context.Context //nolint:containedctx
+	exec   *Executor
+	task   *task
+	result *result
+}
+
+// Error adds errors to the current query, which will be propagated to all
+// queries which depend on it.
+//
+// This will not cause the query to fail; see [Task.Fail] for that.
+func (t *Task) Error(errs ...error) {
+	t.result.Errors = append(t.result.Errors, errs...)
+}
+
+// Fail marks the current query as failed. This will cause all queries that
+// depend on this one to also fail.
+//
+// This function will not return: it will cause a goroutine which calls it to
+// exit. See [Resolve] for notes on what this entails.
+func (t *Task) Fail(errs ...error) {
+	t.Error(errs...)
+	t.result.Failed = true
+	close(t.result.done)
+	runtime.Goexit()
+}
+
+// Resolve executes a set of queries in parallel. Each query is run on its own
+// goroutine.
+//
+// If any of the queries fails, or if the [context.Context] passed to the
+// [Run] call that spawned the [Task] is cancelled, this function calls
+// [runtime.Goexit]. This is not something callers should be concerned about,
+// because every call to a Query execution function is done in its own
+// goroutine, and only occurs when the return value of [Query].Execute will be
+// discarded.
+//
+// Errors for each result are only those that occurred as a direct result of
+// query execution, and *does not* contain that query's transitive errors. This
+// is unlike the behavior of [Run].
+func Resolve[T any](caller Task, queries ...Query[T]) []Result[T] {
+	results := make([]Result[T], len(queries))
+	deps := make([]*task, len(queries))
+	wg := semaphore.NewWeighted(int64(len(queries)))
+
+	for i, q := range queries {
+		i := i
+		deps[i] = caller.exec.getTask(q.URL())
+
+		run(caller, deps[i], q, wg, func() {
+			r := deps[i].result.Load()
+			if r.Value != nil {
+				// This type assertion will always succeed, unless the user has
+				// distinct queries with the same URL, which is a sufficiently
+				// unrecoverable condition that a panic is acceptable.
```

Comment:
While internal panics are okay, they would be categorically bad if they happened in the BSR. Since this is running in a separate goroutine, the calling application (such as BSR code) would have no way of recovering.

The current compiler has a top-level deferred function (in the equivalent of this package's `Run` function, in the `go` function that launches these goroutines) that catches any panic and reports it as a fatal error, specifically for scenarios like this. That way compiler bugs don't turn into serious operational issues.

Reply:
Though I also think it would be better to just return an error here. (Sorry to keep harping on that. But it's much more straight-forward control flow.)

Reply:
I would generally consider this type of panic to be on the same level as a random nil deref. Avoiding panics is not as simple as "never panic"; in this case, it indicates a fairly catastrophic bug on the part of query writers.

How to handle such errors should generally be done by the people who are panic-intolerant, such as with a recover. We can't prevent query authors from panicking out of Execute, after all (unless we try catching that, too).

Reply:
>How to handle such errors should generally be done by the people who are panic-intolerant, such as with a recover. We can't prevent query authors from panicking out of Execute, after all (unless we try catching that, too).

The problem with this is that they can only recover from panics on their (callng) goroutine. If the library function they call spawns a goroutine and the panic happens there (like with this), then there is no ability to catch it and prevent it from crashing the app.

So while I understand that we can't eliminate all panics, since nil-deref/index-out-of-range/etc could still occur, we can _and should_ recover from any panic that happens in a goroutine that is created as an internal detail of this package. The existing compiler in the `protocompile` package does that [here](https://github.com/bufbuild/protocompile/blob/main/compiler.go#L282-L288).